File: 2020/day10/day10.py
#!/usr/bin/env python3
# https://adventofcode.com/2020/day/10

import sys
from collections import defaultdict
from pathlib import Path

verbose = "-v" in sys.argv
if verbose:
    sys.argv.remove("-v")
filename = ("test.txt" if sys.argv[1] == "-t" else sys.argv[1]) if len(sys.argv) > 1 else "input.txt"
data = Path(filename).read_text().strip()


# part 1

adapters = sorted(map(int, data.splitlines()))

diffs = defaultdict(lambda: 0)
for a, b in zip(adapters, adapters[1:]):
    diffs[b - a] += 1

assert len(diffs) == 2
assert (1 in diffs) and (3 in diffs)

diffs[adapters[0] - 0] += 1  # charging outlet has an effective rating of 0 jolts
diffs[3] += 1  # device's built-in adapter is always 3 higher

print(diffs[1] * diffs[3])


# part 2

adapters.insert(0, 0)  # add the charging outlet

n1 = 0
n2 = 0
n3 = 1

for a, b in zip(adapters, adapters[1:]):
    match b - a:
        case 1:
            n1, n2, n3 = n2, n3, n1 + n2 + n3
        case 2:
            n1, n2, n3 = n3, 0, n2 + n3
        case 3:
            n1, n2, n3 = 0, 0, n3

print(n3)

File: 2020/day10/day10.rs
//! [Day 10: Adapter Array](https://adventofcode.com/2020/day/10)

use std::collections::HashMap;

struct Puzzle {
    adapters: Vec<i64>,
}

impl Puzzle {
    const fn new() -> Self {
        Self {
            adapters: Vec::new(),
        }
    }

    /// Get the puzzle input.
    fn configure(&mut self, data: &str) {
        self.adapters
            .extend(data.lines().map_while(|line| line.parse::<i64>().ok()));

        self.adapters.sort_unstable();
    }

    /// Solve part one.
    fn part1(&self) -> u32 {
        let mut diffs: HashMap<i64, u32> = HashMap::new();

        for w in self.adapters.windows(2) {
            let d = w[1] - w[0];

            *diffs.entry(d).or_default() += 1;
        }

        *diffs.entry(self.adapters[0] /*- 0*/).or_default() += 1; // charging outlet has an effective rating of 0 jolts
        *diffs.entry(3).or_default() += 1; // device's built-in adapter is always 3 higher

        diffs[&1] * diffs[&3]
    }

    /// Solve part two.
    fn part2(&self) -> i64 {
        let mut adapters = self.adapters.clone();

        adapters.insert(0, 0); // add the charging outlet

        let mut n = (0, 0, 1);

        for w in adapters.windows(2) {
            n = match w[1] - w[0] {
                1 => (n.1, n.2, n.0 + n.1 + n.2),
                2 => (n.2, 0, n.1 + n.2),
                3 => (0, 0, n.2),
                _ => n,
            }
        }

        n.2
    }
}

fn main() {
    let args = aoc::parse_args();
    let mut puzzle = Puzzle::new();
    puzzle.configure(&args.input);
    println!("{}", puzzle.part1());
    println!("{}", puzzle.part2());
}

/// Test from puzzle input
#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_part1_1() {
        let mut puzzle = Puzzle::new();
        puzzle.configure(&aoc::load_input_data("sample_1.txt"));
        assert_eq!(puzzle.part1(), 7 * 5);
    }

    #[test]
    fn test_part1_2() {
        let mut puzzle = Puzzle::new();
        puzzle.configure(&aoc::load_input_data("sample_2.txt"));
        assert_eq!(puzzle.part1(), 22 * 10);
    }

    #[test]
    fn test_part2() {
        let mut puzzle = Puzzle::new();
        puzzle.configure(&aoc::load_input_data("sample_2.txt"));
        assert_eq!(puzzle.part2(), 19208);
    }
}
